DIPs/DIPxxxx.md

```diff
@@ -0,0 +1,233 @@
+
+```D
+@safe nothrow pure @nogc module someModule;
+
+struct SomeStruct {
+    void someFunction() {
+        // has the desired attributes `@safe nothrow pure @nogc`
+    }
+}
+
+void someTemplateFunction(T...)(T args) {
+    // auto inferred
+    // defaults at the module level do not affect template auto inferring
+    // in contrast to how `@attribute:` does
+}
+
+struct SomeOtherStruct {
+    // @nogc functions...
+
+    @gc void someOtherFunction() {
+        // functions can stay in their logical order
+        // should a function require something other than the default
+        // set at the module level, they can declare so with the
+        // new attribute with the inverse effect
+
+        int[] a = new[10]; // ok
+    }
+
+    // @nogc functions...
+}
+```
+
+Changing the language defaults doesn't address the larger problem. As described in [DIP1028](#Prior&#32;Work) and [DIP10XX](#Prior&#32;Work), which intend to change the language defaults without any easy upgrade path to make old `@system` and `throw` code compilable again. Why `@system:` and `nothrow:` aren't sufficient solutions has already been outlined in the [rationale](#Rationale) regarding the problems with `@attribute:` and defaults. Should `@safe` and `nothrow` become the default, this DIP resolves the problem with letting the user easily choose which defaults they desire for a module. Which before this problem was simply being shifted onto `@system` and `throw` users. With this proposal a user can simply add `@system`, `throw`, `@safe`, or `nothrow` before `module` in each source file and their code will continue to compile, resolving the issue of defaults irregardless of what attributes are actually the default. The default can be easily set to whatever meets a user's requirements for their project.
+
+Ultimately `@attribute:` is more of an antipattern than a feature that should be promoted. When it appears in the middle of a large amount of code it can be difficult to see what the intention is and what behavior is being changed. With this proposal, the default can be set at the module level and the outlying functions can be explicitly marked with clear intention. Otherwise when a function isn't explicitly marked, what attributes a function inherits can be easily found at the very first line of the source file. Ideally `@attribute:` would be deprecated at the module level, but this is not the intention of this proposal nor is it a requirement for any future proposal. This proposal does not serve to dictate the best practices of a user, should they desire to continue to use `@attribute:`, it is their option to.
+
+### Grammer Changes
+
+```diff
+ModuleAttribute:
```

Comment:
The new keywords also need to be added to `StorageClass` so they can be used for `AutoFuncDeclaration`.
```D
pure foo() {} // allowed
impure foo() {} // should be allowed
```
On the other hand, the spec is pretty messy with attributes (see e.g. [issue 20503](https://issues.dlang.org/show_bug.cgi?id=20503)) so I'm not sure how this DIP should deal with that.